MovingBlocks/Terasology
# add IntelliJ configuration

I was working on #3951 and came to the conclusion that it's counterproductive to try to manually tweak gradle and Intellij configs to sync their compiler-output directories.

Wondering “what if I don't do that?” led me to “what if we don't use `./gradlew idea` _at all?_”

So I let IntelliJ import the gradle project and save its project configuration.

What needs to be done to confirm that this is no less broken than the `gradlew idea` version?


## Contains

Support for IntelliJ IDEA 2020.1. Some of it might be backwards-compatible with 2019 versions, but I recommend only trying to support 2020.1+

### An `.idea/` directory

with the project configuration, including features from [the prior `ide.gradle`](https://github.com/MovingBlocks/Terasology/blob/v3.2.0/config/gradle/ide.gradle):

* Checkstyle
* Copyright template
* Terasology-specific annotations marked as entry points
* all the Run Configs.

as well as:

* A startup task that runs `gradlew extractConfigs extractNatives copyInMissingTemplates` so they'll be in place right away.
* Code style setting to not use `*` imports with fewer than 999 imports. (yes, there's really no way to turn it off, setting it to a ridiculously high number is what everyone recommends.)

#### Omitted:

* git stuff. I haven't tested this extensively, but I think recent intellij figures out git roots pretty well on its own?
* compiler "build project automatically" setting. Seems to be stored with the workspace settings, much easier for us to not try to mess with those.


### gitignore changes

* removes some IntelliJ-IDEA ignores, because we want the .idea directory tracked now. The IDE seems to also think you don't want configs tracked if you ignore `.ipr` files, though this uses the directory-based format instead of IPRs.
* Some general clean-up. I think we all have a "somewhat recent git version" compared to that comment from 2015.
* A few things are ignored in `.idea/.gitignore`. I started writing notes in there and it turned in to [a lot of words about how to store .idea in git](https://keturn.net/blog/posts/2020/05/git-add-dot-idea-share-intellij/). I can cross-post that to the forum or wiki or something if that would help.


## How to Test

Things to try, each from a clean checkout (no old build results, no old IntelliJ configs):

* make sure `./gradlew game` works without ever opening it in IntelliJ.
* open in IntelliJ 2020.1
  * agree to the Import Gradle prompt
  * make sure the TerasologyPC run configuration works without ever manually running gradle.
* open in IntelliJ 2020.1
  * agree to the Import Gradle prompt
  * _before_ building, open Settings / Build, Execution, Deployment / Build Tools / Gradle, and set the "Build and run using" option to "Intellij IDEA."
  * Make sure the TerasologyPC run configuration works without ever manually running gradle.

Once you've confirmed that starting from a clean slate works, you'll want to make sure you don't have to strictly commit to one approach or the other. Building in IntelliJ and then later on building from gradle should still work, for example.

To be really thorough, test under all operating systems you want to have people contributing from, and try some variants like:

* on a machine that's never had IntelliJ installed on it before
* on a machine that doesn't have a JDK with JavaFX
* try IntelliJ's "New Project from Version Control" and give it the GitHub URL, instead of checking it out on the command line before opening it.
* say "gradle" backwards three times while spinning counterclockwise
* do your groovy things with the modules and stuff?


## Outstanding Before Merging

* [x] ~review will be less messy if #3955 is merged first~
* …?

## Post-Merge Tasks

- [ ] find all the developer-documentation places that say things like "don't use Import Gradle" or "run gradlew idea" and update or remove them.
